packages/core/src/modules/proofs/protocol/v2/V2ProofService.ts

```diff
@@ -87,21 +88,29 @@ export class V2ProofService<PFs extends ProofFormat[] = ProofFormat[]> extends P
     agentContext: AgentContext,
     options: CreateProposalOptions<PFs>
   ): Promise<{ proofRecord: ProofExchangeRecord; message: AgentMessage }> {
-    const formats = []
+    const attachmentInfo = []
     for (const key of Object.keys(options.proofFormats)) {
       const service = this.formatServiceMap[key]
-      formats.push(
-        await service.createProposal({
-          formats:
-            key === PresentationRecordType.Indy
-              ? await IndyProofUtils.createRequestFromPreview(options)
-              : options.proofFormats,
+
+      let formats
+      if (key === PresentationRecordType.Indy) {
+        const indyFormat = (options as CreateProposalOptions<[IndyProofFormat]>).proofFormats.indy
+        if (!indyFormat) {
+          throw new AriesFrameworkError('No Indy format found.')
+        }
+        formats = await IndyProofUtils.createRequestFromPreview({
+          ...indyFormat,
+          nonce: indyFormat.nonce ?? (await this.wallet.generateNonce()),
         })
```

Comment:
Maybe we can move the indy createRequestFromPreview logic to the indy proof format service? I know you didn't write this, but we should aim to leave format specific code out of the format service and it will make for more beatiful code I think

Reply:
You are absolutely right! After analyzing it a bit more I think the same behaviour can be respected without that `if (key === PresentationRecordType.Indy)`, and just calling IndyProofUtils.createRequestFromPreview from `IndyProofFormatService`. 

Actually I think there is room for improvement but I didn't want to touch too much because I'm not so familiar with proofs code. At least `V2ProofService` seems cleaner now :smile: .